torch/fx/experimental/symbolic_shapes.py

```diff
@@ -2115,9 +2146,8 @@ def produce_guards(
         dynamic_sources = []
         constraint_violations : List[Tuple[Constraint, Callable[[], str]]] = []
 
-        def record_constraint_violation(constraint, msg_fn):
-            assert callable(msg_fn)
-            constraint_violations.append((constraint, msg_fn))
+        def record_constraint_violation(constraint, msg):
+            constraint_violations.append((constraint, lambda: msg))
```

Comment:
moving the lambda in here is missing the point, the point is we need to delay the evaluation of msg

Reply:
The way it was before, we were capturing variables by name that could change in value by the time the error is raised. This caused a lot of confusion for me. I had to meticulously rename captured variables in each message to ensure it accurately reflected the state when the lambda was created.

If we want to delay the evaluation of the message (what's so heavy about them?), we could pass in *args to the lambda.

Reply:
Oh, yes, that's a legit bug!

The reason the messages need to be delayed is because the `hint()` call refers to `symbol_to_source`, which is still being built at the time we register constraints, in the loop here:

```
        def track_symint(source, val, constraint=None):
            if isinstance(val, SymInt):
                s = val.node.expr
        
                if isinstance(s, sympy.Symbol):
                    symbol_to_source[s].append(source)
                    if constraint is not None:
                        symbol_to_constraints[s].append(constraint)
                elif isinstance(-s, sympy.Symbol):
                    symbol_to_source[-s].append(NegateSource(source))
                else:
                    if constraint is not None:
                        # TODO: Maybe non-strict constraint shouldn't error
                        # here?  Check what happens in practice
                        def hint():
                            if s.free_symbols:
                                return (
                                    "Perhaps you meant to specify a constraint on {s.free_symbols}?" +
                                    "; ".join(
                                        f"{s0} bound by " + ", ".join(str(source0) for source0 in symbol_to_source[s0])
                                        for s0 in s.free_symbols
                                    )
                                )
                            else:
                                return "Did you really mean to mark this dimension as dynamic?"
            
                        record_constraint_violation(constraint, lambda: (
                            f"Could not validate constraint {constraint.render(source)} as "
                            f"{source.name()} is actually a non-atomic symbolic expression "
                            f"{s}.  {hint()}"
                        ))
```

I suppose you could index `symbol_to_source` FIRST, and then dispense with the lambdas. The main point is that some of the error messages rely on context which may not be entirely constructed at the time we produce the error.

Reply:
I ended up limiting the use of capture to just the `hint` and using `functools.partial` to explicitly capture the variable `s` in the loop, which I believe was also getting captured incorrectly (it would have likely taken the last value out of the loop, given the hint would be called after the loop).